### PR TITLE
Migrate to Qt6

### PR DESCRIPTION
**Pull request by dyfer**
## Purpose and Motivation

Migrate to Qt6!
===========

- Supported Qt version: 6.2 and up
  - 5.15 still works, open for discussion
  - macOS and Windows builds use official Qt 6.7.2; 6.8.0 should come out at the end of September, but using it would raise the deployment target on macOS even higher (targeted platforms to be discussed in #6482)
- macOS build seems to have a visual issue with the size of the icons in menus; I'm not sure why that is. Windows looks fine
- when using Qt from homebrew, deployment needs to be fixed for ~~a number of libraries~~ `QtQuickWidgets`
- `verify_app` doesn't seem to work with neither the official Qt6 (a known issue), nor with Qt6 from homebrew (_new_ issue); details in #6481
- QuartzComposerView is still present and builds when using Qt5, i.e. the macOS legacy build still has it

----
Documentation updates will be submitted in #6483
Thanks @jrsurge for starting that work back in the day
Thanks @xunil-cloud for fixing the WebView migration

## Types of changes

- New feature (?)

## To-do list

- [x] Code is tested
- [x] All tests are passing
- n/a Updated documentation - #6483
- [x] This PR is ready for review

### CODE REVIEW

**Review by capital-G — COMMENTED** [review]
Wow, thanks so much for this! This is a really great achievement and contribution @dyfer ! 

image

❤️ 

Also the icons are really neat

image

As mentioned, I can verify the mentioned icon bug, e.g. the disk icon has grown a bit. Maybe this is a hidapi thing (uneducated guess^^)?

image

Do you have a tendency if we should focus on having Qt6 merged and fix the bugs from within dev? I think I'd prefer to have it this way, as it allows earlier testing, better distribution of bug fixing and also lowers the danger of growing stale. And it is a way we want to go nonetheless.

Maybe you can add some insights why QuartzComposer is available in Qt5 but not in Qt6 anymore? I am not too familiar with it r/n and maybe it is something obvious, but maybe leave a comment somewhere?
So in case Qt5 gets removed we know why we could/could not remove it?

On a sidenote: Configuring the build takes longer on macOS. Took around 10 secs, now 60. Also a "cold build" takes now 4 instead of 3 minutes, but thats life.

Also there are some CLI logs while using it, it definitely got a bit noisier compared to Qt5

```
objc[23375]: Class QT_ROOT_LEVEL_POOL__THESE_OBJECTS_WILL_BE_RELEASED_WHEN_QAPP_GOES_OUT_OF_SCOPE is implemented in both /Users/scheiba/github/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtCore.framework/Versions/A/QtCore (0x10c228608) and /usr/local/Cellar/qt/6.7.0_2/lib/QtCore.framework/Versions/A/QtCore (0x12e8fd608). One of the two will be used. Which one is undefined.
objc[23375]: Class KeyValueObserver is implemented in both /Users/scheiba/github/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtCore.framework/Versions/A/QtCore (0x10c228630) and /usr/local/Cellar/qt/6.7.0_2/lib/QtCore.framework/Versions/A/QtCore (0x12e8fd630). One of the two will be used. Which one is undefined.
objc[23375]: Class RunLoopModeTracker is implemented in both /Users/scheiba/github/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtCore.framework/Versions/A/QtCore (0x10c228680) and /usr/local/Cellar/qt/6.7.0_2/lib/QtCore.framework/Versions/A/QtCore (0x12e8fd680). One of the two will be used. Which one is undefined.
objc[23375]: Class QDarwinPermissionHandler is implemented in both /Users/scheiba/github/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtCore.framework/Versions/A/QtCore (0x10c2286f8) and /usr/local/Cellar/qt/6.7.0_2/lib/QtCore.framework/Versions/A/QtCore (0x12e8fd6f8). One of the two will be used. Which one is undefined.
Path override failed for key base::DIR_APP_DICTIONARIES and path '/Users/scheiba/github/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/MacOS/qtwebengine_dictionaries'
Path override failed for key base::DIR_APP_DICTIONARIES and path '/Users/scheiba/github/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Versions/A/Helpers/QtWebEngineProcess.app/Contents/MacOS/qtwebengine_dictionaries'
qt.pointer.dispatch: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=1961.73,606.96 gbl=1961.73,606.96 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-1961.73,-606.96 last=-1961.73,-606.96 Δ 1961.73,606.96) : no target window
qt.pointer.dispatch: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=1763.4,531.82 gbl=1763.4,531.82 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-1763.4,-531.82 last=-1763.4,-531.82 Δ 1763.4,531.82) : no target window
qt.pointer.dispatch: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=1601,664.216 gbl=1601,664.216 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-1601,-664.216 last=-1601,-664.216 Δ 1601,664.216) : no target window
qt.pointer.dispatch: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=1568.28,524.186 gbl=1568.28,524.186 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-1568.28,-524.186 last=-1568.28,-524.186 Δ 1568.28,524.186) : no target window
qt.pointer.dispatch: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=1695.82,609.047 gbl=1695.82,609.047 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-1695.82,-609.047 last=-1695.82,-609.047 Δ 1695.82,609.047) : no target window
qt.pointer.dispatch: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=1457.29,1019.74 gbl=1457.29,1019.74 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-1457.29,-1019.74 last=-1457.29,-1019.74 Δ 1457.29,1019.74) : no target window
qt.pointer.dispatch: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=1680.34,474.557 gbl=1680.34,474.557 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-1680.34,-474.557 last=-1680.34,-474.557 Δ 1680.34,474.557) : no target window
```

Also, I found that macOS using qt5 compiles but it does not start?

```
❯ cmake -G Xcode -DCMAKE_PREFIX_PATH=`brew --prefix qt@5`  ..
❯ make install...
... success...
❯ build/Install/SuperCollider/SuperCollider.app/Contents/MacOS/SuperCollider
qt.qpa.plugin: Could not find the Qt platform plugin "cocoa" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

[1]    32608 abort      build/Install/SuperCollider/SuperCollider.app/Contents/MacOS/SuperCollider
```

Someone with some CMake expertise should take a look at this, I can only try to understand what's going on but not evaluate if it is the best way to do things.

The C++ changes look well and reasonable. The Qt5/6 dual support is not as *ugly* as I would have thought it would be, therefore I may revert my mentioned initial opposition in #6482  towards Qt5/6 dual support.

**Comment by dyfer** [comment]
> Also the icons are really neat

I wish I could take credit - they just showed up when building with Qt6.


> As mentioned, I can verify the mentioned icon bug, e.g. the disk icon has grown a bit. Maybe this is a hidapi thing (uneducated guess^^)?

HiDPI should work as it has in Qt 5.15, AFAIU. There might be something specific in our codebase (i.e. setting sizes of stuff), but the basic setup for the HiDPI should be fine... I.e. I have no idea what the issue is, but it's unlikely just some global setting we can change.


> Do you have a tendency if we should focus on having Qt6 merged and fix the bugs from within dev? I think I'd prefer to have it this way, as it allows earlier testing, better distribution of bug fixing and also lowers the danger of growing stale. And qt6 is the path we want to go nonetheless.

I wouldn't mind if a few pairs of eyes look at this PR before we merge it.

> 
> Maybe you can add some insights why QuartzComposer is available in Qt5 but not in Qt6 anymore? 

QuartzCompose has been removed from Qt6. It's just an older macOS technology that Qt dropped. We deprecated it in SC (#5710) and it's mentioned in the help, do we need anything else?

> On a sidenote: Configuring the build takes longer on macOS. Took around 10 secs, now 60. Also a "cold build" takes now 4 instead of 3 minutes, but thats life.

Yeah, I think that's just how it goes.

> Also there are some CLI logs while using it, it definitely got a bit noisier compared to Qt5
> 
> ```
> objc[23375]: Class QT_ROOT_LEVEL_POOL__THESE_OBJECTS_WILL_BE_RELEASED_WHEN_QAPP_GOES_OUT_OF_SCOPE is implemented in both /Users/scheiba/github/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtCore.framework/Versions/A/QtCore (0x10c228608) and /usr/local/Cellar/qt/6.7.0_2/lib/QtCore.framework/Versions/A/QtCore (0x12e8fd608). One of the two will be used. Which one is undefined.
> objc[23375]: Class KeyValueObserver is implemented in both /Users/scheiba/github/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtCore.framework/Versions/A/QtCore (0x10c228630) and /usr/local/Cellar/qt/6.7.0_2/lib/QtCore.framework/Versions/A/QtCore (0x12e8fd630). One of the two will be used. Which one is undefined.
> objc[23375]: Class RunLoopModeTracker is implemented in both /Users/scheiba/github/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtCore.framework/Versions/A/QtCore (0x10c228680) and /usr/local/Cellar/qt/6.7.0_2/lib/QtCore.framework/Versions/A/QtCore (0x12e8fd680). One of the two will be used. Which one is undefined.
> objc[23375]: Class QDarwinPermissionHandler is implemented in both /Users/scheiba/github/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtCore.framework/Versions/A/QtCore (0x10c2286f8) and /usr/local/Cellar/qt/6.7.0_2/lib/QtCore.framework/Versions/A/QtCore (0x12e8fd6f8). One of the two will be used. Which one is undefined.
>```

IIUC that happens always when you have bundled Qt in an app and Qt installed in a system location. This did not happen with Qt5 because that wouldn't be linked in the system location by homebrew by default. This also came up when discussing "new" standalone app approach on macOS a while back. I don't think we can do anything about it.

>```
> Path override failed for key base::DIR_APP_DICTIONARIES and path '/Users/scheiba/github/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/MacOS/qtwebengine_dictionaries'
> Path override failed for key base::DIR_APP_DICTIONARIES and path '/Users/scheiba/github/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Versions/A/Helpers/QtWebEngineProcess.app/Contents/MacOS/qtwebengine_dictionaries'
> qt.pointer.dispatch: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=1961.73,606.96 gbl=1961.73,606.96 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-1961.73,-606.96 last=-1961.73,-606.96 Δ 1961.73,606.96) : no target window
> qt.pointer.dispatch: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=1763.4,531.82 gbl=1763.4,531.82 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-1763.4,-531.82 last=-1763.4,-531.82 Δ 1763.4,531.82) : no target window
> qt.pointer.dispatch: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=1601,664.216 gbl=1601,664.216 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-1601,-664.216 last=-1601,-664.216 Δ 1601,664.216) : no target window
> qt.pointer.dispatch: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=1568.28,524.186 gbl=1568.28,524.186 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-1568.28,-524.186 last=-1568.28,-524.186 Δ 1568.28,524.186) : no target window
> qt.pointer.dispatch: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=1695.82,609.047 gbl=1695.82,609.047 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-1695.82,-609.047 last=-1695.82,-609.047 Δ 1695.82,609.047) : no target window
> qt.pointer.dispatch: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=1457.29,1019.74 gbl=1457.29,1019.74 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-1457.29,-1019.74 last=-1457.29,-1019.74 Δ 1457.29,1019.74) : no target window
> qt.pointer.dispatch: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=1680.34,474.557 gbl=1680.34,474.557 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-1680.34,-474.557 last=-1680.34,-474.557 Δ 1680.34,474.557) : no target window
> ```


These might be some legitimate issues to look into. @capital-G did you notice that the IDE is a bit more sluggish in interaction?


> > Also, I found that macOS using qt5 compiles but it does not start?

I found that you need to unlink (or uninstall, but unlink is easier) qt6 before building with qt5. `brew unlink qt6`.


> 
> The C++ changes look well and reasonable. The Qt5/6 dual support is not as _ugly_ as I would have thought it would be, therefore I may revert my mentioned initial opposition in #6482 towards Qt5/6 dual support.

<3

File: editors/sc-ide/CMakeLists.txt
Comment (dyfer) [reply editors/sc-ide/CMakeLists.txt:19]: I don't remember anymore :) But it works right, it seems.

File: editors/sc-ide/CMakeLists.txt
Comment (dyfer) [reply editors/sc-ide/CMakeLists.txt:512]: No ticket, as this came up only in Qt6. These issues come up occasionally and it's typically finger pointing between Qt and homebrew. I found it in homebrew Qt6 build so I patched it... not sure if there's a better way to go about this.

File: .github/workflows/build_macos.yml
Comment (dyfer) [reply .github/workflows/build_macos.yml:172]: Yeah, I wasn't sure anymore which one is more proper. I'm happy to change that.

File: QtCollider/CMakeLists.txt
Comment (dyfer) [reply QtCollider/CMakeLists.txt:21]: We are printing later in the configuration the version of the Qt that's found. Do you think we should print extra in here?

File: lang/CMakeLists.txt
Comment (dyfer) [reply lang/CMakeLists.txt:211]: I think I just kept the tabs here, as these lines already used them, but I can change this to spaces.

**Comment by JordanHendersonMusic** [comment]
This is excellent, thank you!
 
I've got a question...
What happens if in the future we use a feature only available in qt6? Will we be dropping qt5 support? Or are we going to limit ourselves to just their intersection?

**Comment by dyfer** [comment]
> What happens if in the future we use a feature only available in qt6? Will we be dropping qt5 support? Or are we going to limit ourselves to just their intersection?

Eventually we'll definitely drop Qt 5.15 altogether. There was some discussion in #6482 if we should do this already here or not. If we keep Qt 5.15 support here, I'd treat it as present "for the time being", possibly no longer than 3.14 release

If the Qt6-only feature comes up before we drop Qt 5.15, then we'd decide if we drop Qt 5.15 at that time, or use preprocessor switches to have this feature only in Qt6 builds.

File: lang/CMakeLists.txt
Comment (capital-G) [reply lang/CMakeLists.txt:211]: On my side it looks a bit strange in context, but this was also the case before this PR. But while we are at it, we could make it consistent.

image

File: QtCollider/CMakeLists.txt
Comment (capital-G) [reply QtCollider/CMakeLists.txt:21]: You're right - I think this will be sufficient.

File: .github/workflows/build_macos.yml
Comment (capital-G) [reply .github/workflows/build_macos.yml:172]: I think the canonical way within brew for versions seems to be via `@`, which may even allow finer grained versioning? Would actually prefer to switch it in this case.

File: editors/sc-ide/CMakeLists.txt
Comment (capital-G) [reply editors/sc-ide/CMakeLists.txt:512]: Maybe just leave a link to the relevant homebrew Qt6 build/patch here so in case this breaks/changes we can take a look how their code developed?

Comment (dyfer) [reply editors/sc-ide/CMakeLists.txt:512]: Sorry, I meant that I found it while building SC with homebrew Qt6. There's no "patch" on homebrew/qt end, I was referring to my Cmake entry as the "patch". Macdeployqt sometimes has bugs that come out in homebrew builds, but not in the official Qt release. 
I know it would be good to report it, but that would need a clean/small reproducer, which I don't have atm.

File: lang/CMakeLists.txt
Comment (dyfer) [reply lang/CMakeLists.txt:211]: Sounds good, I'll fix this.

File: editors/sc-ide/CMakeLists.txt
Comment (dyfer) [reply editors/sc-ide/CMakeLists.txt:512]: I think it might be desirable to try to get to the bottom of this, but I'm not sure if I'll be able to get to it for the time being. Should that block this PR or not?

File: lang/CMakeLists.txt
Comment (dyfer) [reply lang/CMakeLists.txt]: Fixed

File: .github/workflows/build_macos.yml
Comment (dyfer) [reply .github/workflows/build_macos.yml]: done

File: editors/sc-ide/CMakeLists.txt
Comment (capital-G) [reply editors/sc-ide/CMakeLists.txt:512]: No, definitely not a blocker at all. Just if you would have some resources would have been a good place to put them there, but since there are none, there is nothing to put there.

**Comment by capital-G** [comment]
@dyfer What are your plans for merging this? Is it ready "as-is", is something missing? I looked over it, the C++ code looks fine to me (but I am not an expert in this domain), and I didn't delved into the CMake files too much.

Personally, if another person has a look over this I would like to have this merged rather sooner than later.

**Comment by dyfer** [comment]
I think this can be merged as is. I'd love the little issues with appearance to be ironed out eventually, but I don't have the bandwidth to do this myself, and I think the build is usable as is.

We'd keep 5.15 compatibility as is so we can even temporarily revert builds to 5.15 if some major issue comes up. But having this in develop will likely get more attention to it which would likely help ironing out any issues.

**Comment by capital-G** [comment]
@JordanHendersonMusic do you want to give this a review? :)

**Review by capital-G — APPROVED** [review]

**Comment by mtmccrea** [comment]
I don't have enough familiarity to review this, but knowing that much attention and care has gone into this, I agree with @dyfer about this sensible next step 

> But having this in develop will likely get more attention to it which would likely help ironing out any issues.

**Comment by capital-G** [comment]
Maybe we should therefore ask - are there any objections to this getting merged?

Maybe give it a grace period of one week (so 2024-10-13) and if there is no additional review we will merge it?

**Comment by capital-G** [comment]
Just once more, many thanks for taking care of this @dyfer , I think we all can be very glad about this contribution and will put SC on a good track for the next years!

**Pull request merged**
